File: NWorldPermissions/src/top/nololiyt/worldpermissions/commands/marks/AddExecutor.java
package top.nololiyt.worldpermissions.commands.marks;

import org.bukkit.Location;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import top.nololiyt.worldpermissions.configurationmanagers.LocalMarksManager;
import top.nololiyt.worldpermissions.RootPlugin;
import top.nololiyt.worldpermissions.entitiesandtools.DotDividedStringBuilder;
import top.nololiyt.worldpermissions.entitiesandtools.MessagesSender;
import top.nololiyt.worldpermissions.entitiesandtools.StringPair;
import top.nololiyt.worldpermissions.commands.Executor;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class AddExecutor extends Executor
{
    protected final static String layerName = "add";
    
    @Override
    public String permissionName()
    {
        return null;
    }
    
    @Override
    public String messageKey()
    {
        return layerName;
    }
    
    @Override
    public List<String> getTabComplete(RootPlugin rootPlugin,int ordinal)
    {
        switch (ordinal)
        {
            default:
                return new ArrayList<>();
        }
    }
    
    @Override
    protected boolean run(int layer, RootPlugin rootPlugin,
                          DotDividedStringBuilder messageKey, CommandSender commandSender,
                          String[] args)
    {
        MessagesSender messagesSender = new MessagesSender(rootPlugin.getMessagesManager(),
                commandSender, null);
    
        if (!(commandSender instanceof Player))
        {
            messagesSender.setArgs(new StringPair[]{
                    StringPair.senderName(commandSender.getName()),
            });
            messagesSender.send(messageKey.append("without-a-position"));
            return true;
        }
    
        if (args.length - 1 != layer)
            return false;
    
        Player player = (Player) commandSender;
    
        String markName = args[layer];
    
        messagesSender.setArgs(new StringPair[]{
                StringPair.markName(markName),
                StringPair.senderName(player.getDisplayName())
        });
        try
        {
            addMark(rootPlugin.getLocalMarksManager(), markName,
                    player.getLocation(), messagesSender,
                    messageKey);
        }
        catch (IOException ex)
        {
            ex.printStackTrace();
            messagesSender.send(messageKey.append("failed"));
        }
        return true;
    }
    
    private void addMark(LocalMarksManager localMarksManager, String markName, Location location,
                         MessagesSender messagesSender, DotDividedStringBuilder messageKey)
            throws IOException
    {
        if (localMarksManager.getMark(markName) != null)
        {
            messagesSender.send(messageKey.append("with-occupied-name"));
            return;
        }
    
        localMarksManager.setMark(markName, location);
        messagesSender.send(messageKey.append("completed"));
    }
}

File: NWorldPermissions/src/top/nololiyt/worldpermissions/commands/marks/AddExecutor.kt
package top.nololiyt.worldpermissions.commands.marks

import org.bukkit.command.CommandSender
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player
import top.nololiyt.worldpermissions.MarksManager
import top.nololiyt.worldpermissions.RootPlugin
import top.nololiyt.worldpermissions.entities.DotDividedStringBuilder
import top.nololiyt.worldpermissions.entities.StringPair
import top.nololiyt.worldpermissions.commands.Executor

import java.io.File
import java.io.IOException

class AddExecutor : Executor()
{
    
    override fun permissionName(): String?
    {
        return null
    }
    
    override fun messageKey(): String?
    {
        return layerName
    }
    
    override fun run(layer: Int, rootPlugin: RootPlugin, permission: DotDividedStringBuilder,
        messageKey: DotDividedStringBuilder, commandSender: CommandSender, args: Array<String>): Boolean
    {
        if (commandSender !is Player)
        {
            val pairs = arrayOf<StringPair?>(StringPair.senderName(commandSender.name))
            
            rootPlugin.messagesManager.sendMessage(messageKey.append("without-a-position"), pairs, commandSender);
            return true
        }
        
        if (args.size - 1 != layer) return false
        
        val cPairs =
            arrayOf<StringPair?>(StringPair.markName(args[layer]), StringPair.senderName(commandSender.displayName))
        try
        {
            val marksManager = rootPlugin.marksManager
            if (marksManager!!.getMark(args[layer]) != null)
            {
                rootPlugin.messagesManager.sendMessage(messageKey.append("with-occupied-name"), cPairs, commandSender);
                return true
            }
            marksManager.setMark(args[layer], commandSender.location)
            
            rootPlugin.messagesManager.sendMessage(messageKey.append("completed"), cPairs, commandSender);
            return true
        }
        catch (ex: IOException)
        {
            ex.printStackTrace()
            
            rootPlugin.messagesManager.sendMessage(messageKey.append("failed"), cPairs, commandSender);
            return true
        }
        
    }
    
    companion object
    {
        protected val layerName = "add"
    }
}